scylla/src/transport/connection.rs

```diff
@@ -312,7 +313,7 @@ impl Connection {
         &self,
         prepared_statement: &PreparedStatement,
         values: impl ValueList,
-        paging_state: Option<Bytes>,
+        paging_state: &Option<Bytes>,
```

Comment:
I wish I could, but it's called from within a lambda which implements Fn (and not FnOnce), and making paging_state a reference here was the least painful way of dealing with the borrow checker I thought of (also, same trick is applied to values nearby). Suggestions welcome :)

Reply:
Here you go:

```diff
diff --git a/scylla/src/transport/connection.rs b/scylla/src/transport/connection.rs
index f256b3a..a14ca0a 100644
--- a/scylla/src/transport/connection.rs
+++ b/scylla/src/transport/connection.rs
@@ -305,7 +305,7 @@ impl Connection {
         &self,
         prepared_statement: &PreparedStatement,
         values: impl ValueList,
-        paging_state: &Option<Bytes>,
+        paging_state: Option<Bytes>,
     ) -> Result<QueryResult, QueryError> {
         self.execute(prepared_statement, values, paging_state)
             .await?
@@ -316,7 +316,7 @@ impl Connection {
         &self,
         prepared_statement: &PreparedStatement,
         values: impl ValueList,
-        paging_state: &Option<Bytes>,
+        paging_state: Option<Bytes>,
     ) -> Result<QueryResponse, QueryError> {
         let serialized_values = values.serialized()?;
 
@@ -327,7 +327,7 @@ impl Connection {
                 serial_consistency: prepared_statement.get_serial_consistency(),
                 values: &serialized_values,
                 page_size: prepared_statement.get_page_size(),
-                paging_state: paging_state.clone(),
+                paging_state,
             },
         };
 
diff --git a/scylla/src/transport/iterator.rs b/scylla/src/transport/iterator.rs
index ac84a42..4de0805 100644
--- a/scylla/src/transport/iterator.rs
+++ b/scylla/src/transport/iterator.rs
@@ -159,7 +159,7 @@ impl RowIterator {
 
             let page_query = |connection: Arc<Connection>, paging_state: Option<Bytes>| async move {
                 connection
-                    .execute(prepared_ref, values_ref, &paging_state)
+                    .execute(prepared_ref, values_ref, paging_state)
                     .await
             };
 
diff --git a/scylla/src/transport/session.rs b/scylla/src/transport/session.rs
index 6f58758..7b0189a 100644
--- a/scylla/src/transport/session.rs
+++ b/scylla/src/transport/session.rs
@@ -465,7 +465,7 @@ impl Session {
             |node: Arc<Node>| async move { node.connection_for_token(token).await },
             |connection: Arc<Connection>| async move {
                 connection
-                    .execute_single_page(prepared, values_ref, paging_state_ref)
+                    .execute_single_page(prepared, values_ref, paging_state_ref.clone())
                     .await
             },
         )
```

I think the lambda you wrote about is at the bottom of the diff - explicitly cloning `paging_state_ref` does the trick. In case of the lambda, cloning is necessary because it can be called multiple times, so this diff does not change the number of clones made there.

On the other hand, look at the diff in `RowIterator` - in that case there will be one less clone because the paging state will be passed by value and `execute` won't have to clone it.

Reply:
:facepalm: right... I was almost there with paging_state_ref and gave up, while the only thing left to make it work was to `clone()` in lambda's body. Thx, will do in v2